### PR TITLE
Feature/joining to communities

### PR DESCRIPTION
**Pull request by konrad-paszek**
Add "join to communities" logic
https://app.clickup.com/t/8694r3z1g

## Summary by Sourcery

Add functionality for users to join communities, including backend logic and frontend integration. Refactor community detail handling to improve code reuse and maintainability. Update tests to cover new joining functionality.

New Features:
- Implement the ability for users to join communities, including a new view and form handling for joining.

Enhancements:
- Refactor community detail view to use a mixin for retrieving community objects, improving code reuse.

Tests:
- Add tests for the new community joining feature, ensuring users can join communities and receive appropriate messages.

### CODE REVIEW

File: core/views.py [hunk 1]
```diff
@@ -164,6 +164,18 @@ def post(self: "PostSaveView", request: HttpRequest, pk: int, action_type: str)
         return redirect("post-detail", pk=post.id)
 
 
+class CommunityMixin:
+    model = Community
+
+    def get_object(self: "CommunityJoin") -> Community:
```
[the remark this answers is not in the document]
Comment (konrad-paszek) [reply core/views.py:170]: Joining should be available from every level, logic to be approved by moderator will be added under other ticket

File: core/tests/test_views.py [hunk 2]
```diff
@@ -452,6 +452,31 @@ def test_create_community_view(client: Client, user: User) -> None:
     )
 
 
+def test_join_community_view(client: Client, user: User, community: Community) -> None:
```
Comment (jacoor) [inline core/tests/test_views.py]: This test seems too big. 
Test should have 3 sections
- Arrange
- Action
- Assert

I see several Arrange Action and Assert sections here. 
Please split into separate unit tests with correct namings

File: core/tests/test_views.py [hunk 3]
```diff
@@ -452,6 +452,31 @@ def test_create_community_view(client: Client, user: User) -> None:
...
+    url = reverse("community-join", kwargs={"slug": "not_exist"})
+    response = client.post(url)
+    assert response.status_code == 404
+    assert "Community does not exist" in str(response.context)
+
+    client.alogout()
+    url = reverse("community-join", kwargs={"slug": community.slug})
+    response = client.post(url)
+    assert response.status_code == 302
+    assert "You have joined the community!" not in str(response.context)
```
Comment (jacoor) [inline core/tests/test_views.py]: Yea, this is misleading. Having this in a separate test like "is_not_logged_in_is_unable_to_join_community" would be much clearer. I got confused. 
Also, a good test would be to show that this user was redirected to a login page or that there was an error message. This shows there was no success, which is hard to wrap my head around.

File: core/tests/test_views.py [hunk 4]
```diff
@@ -452,6 +452,31 @@ def test_create_community_view(client: Client, user: User) -> None:
...
+    messages = list(get_messages(response.wsgi_request))
+    assert response.status_code == 302
+    assert any("You have joined the community!" in message.message for message in messages)
+
+    response = client.post(url)
+    messages = list(get_messages(response.wsgi_request))
+    assert any("You are already a member of this community." in message.message for message in messages)
+    assert response.status_code == 302
+
+    url = reverse("community-join", kwargs={"slug": "not_exist"})
```
Comment (jacoor) [inline core/tests/test_views.py:482]: separate test.. Another AAA sections

File: core/views.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (jacoor) [reply core/views.py:170]: @konrad-paszek AI is right. 
At least please add a check and throw an expection that it has to be implemented
ie raise this one: https://docs.python.org/3/library/exceptions.html#NotImplementedError
The problem is, leaving this hole open is asking for trouble - we may miss it in testing and in later implementation. 
Do not trust future developers that they will implement everything that is in backlog. And fail loudly. It's better to have a bug then to have privacy issue (in this exact scenario).

**Review by jacoor — CHANGES_REQUESTED** [review]

**Review by jacoor — APPROVED** [review]

**Pull request merged**